homeassistant/components/point/config_flow.py

```diff
@@ -0,0 +1,131 @@
+"""Config flow for Minut Point."""
+from collections import OrderedDict
+
+import logging
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.core import callback
+
+
+CLIENT_ID = 'client_id'
+CLIENT_SECRET = 'client_secret'
+KEY_IMPORT_GROUPS = 'import_groups'
+
+AUTH_CALLBACK_PATH = '/api/minut'
+AUTH_CALLBACK_NAME = 'api:minut'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AuthError(Exception):
+    """Exception if authentication occurs."""
+
+    def __init__(self, code):
+        """Initialize exception."""
+        super().__init__()
+        self.code = code
+
+
+@config_entries.HANDLERS.register('point')
+class FlowHandler(config_entries.ConfigFlow):
+    """Handle a config flow."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize flow."""
+        self._client_id = None
+        self._client_secret = None
+        self._token = None
+        self._point_session = None
+        self._redirect_uri = None
+
+    async def token(self, code):
+        """Register token."""
+        token = self._point_session.get_access_token(code)
+        _LOGGER.debug("Got new token: %s", token)
+        if not self._point_session.is_authorized:
+            _LOGGER.error('Authentication Error')
+        else:
+            _LOGGER.info('Successfully authenticated Point')
+            self._token = token
+
+    async def async_step_user(self, user_input=None):
+        """Handle the authentication with a gateway."""
+        errors = {}
+
+        if user_input is not None:
+            self._client_id = user_input.get(
+                CLIENT_ID, self._client_id)
+            self._client_secret = user_input.get(
+                CLIENT_SECRET, self._client_secret)
+            return await self.async_step_auth()
+
+        fields = OrderedDict()
+        fields[vol.Required(CLIENT_ID, default='')] = str
```

Comment:
We should not ask for client ID and secret in config flows. They should be configured via `configuration.yaml`, after which we can do a normal config flow.

Reply:
Why is that?

I was super happy to get a nice integration flow without requiring the user to fiddle in configuration.yaml.

I have actually spent a few days getting everything working with the entity registration and the integration flow so hearing that this is not the "right way" gets me quite annoyed and frustrated.

Reply:
With config entries we aim on getting a UI working for anyone, everywhere. Users should not have to register on a developer portal to get a client ID and secret, they should just start a config flow, log in and be done. We're doing it right this time.

Client ID and secret will need to be set in configuration.yaml. See the Nest component as an example on how to implement this: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/nest/config_flow.py#L23

In the future, the user will not have to set a client ID or secret at all as we can register for credentials with Home Assistant Cloud and provide this to the users (for free obviously). To see how this will look, see [this branch](https://github.com/home-assistant/home-assistant/commit/5efbe119f974d3e12a8ce697a456a9a1bcd18f52).

And no, we don't want to accept Client ID / Secret now and remove it in the future once all other pieces are in place, because it means that in the future we will have to do a lot of work to remove things to get to the experience that we want.

Reply:
Ok! 

I'll try to get it working like the Nest integration then. I've been reading the source for Nest for a while but it is quite messy. Will I still have access to device-registry? 

Am I allowed to do this (save the tokens in the config_registry) or do I have to use an external `json`?

```python
    def token_saver(token):
        _LOGGER.debug('Saving updated token')
        entry.data[CONF_TOKEN] = token
        hass.config_entries.async_update_entry(
            entry, data={**entry.data})
```



What is the timeframe for Cloud integration?

Reply:
yeah saving the token to config entry = 👍 that's what is supposed to be done. 

Nest has an oauth flow that supports "pin" where a user copies a pin into the config flow. Having a webview is ok .

Reply:
I've been reading the Nest integration now and don't understand why `register_flow_implementation()` takes in `domain` and the `len(flows) == 1:` in `config_flow.py` it seems to me like that the length is always 1 or am I missing something?

Reply:
It's build so that multiple config flow implementations can be configured, as the `cloud` component will provide one in the future.

Reply:
> It's build so that multiple config flow implementations can be configured, as the `cloud` component will provide one in the future.

Perfect! I've adopted the code according to the Nest integration now. 

The only thing I didn't manage is to close the flow as I'm side tracking it with a `WebView` and create entries when I receive the code back. I had to do a

```python
            return self.async_abort(reason='external_setup')
```
Where `'external_setup'` is a success message.

Reply:
That's fine for now. I have a branch to actually fix this properly with a new [external step](https://github.com/home-assistant/home-assistant/blob/nest-account-link/homeassistant/data_entry_flow.py#L193-L202). However, that branch is from April and kinda stale.